File: web/src/lib/features.ts
// Technical indicator feature engineering — TypeScript port of modal/features.py

export type OHLCVRecord = {
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
};

export type FeatureRow = {
  sma_ratio: number;
  price_to_sma20: number;
  rsi: number;
  rsi_change: number;
  macd_hist: number;
  bb_position: number;
  atr_ratio: number;
  volume_ratio: number;
  obv_change: number;
  price_change: number;
  price_change_5d: number;
  high_low_range: number;
  trend_strength: number;
  vol_regime: number;
  mean_rev_dist: number;
  rsi_zone: number;
  consecutive_up: number;
};

export const FEATURES = [
  "sma_ratio",
  "price_to_sma20",
  "rsi",
  "rsi_change",
  "macd_hist",
  "bb_position",
  "atr_ratio",
  "volume_ratio",
  "obv_change",
  "price_change",
  "price_change_5d",
  "high_low_range",
  "trend_strength",
  "vol_regime",
  "mean_rev_dist",
  "rsi_zone",
  "consecutive_up",
] as const;

// ── Helpers ───────────────────────────────────────────────────────────────────

function sma(values: number[], window: number): number[] {
  const result = new Array<number>(values.length).fill(NaN);
  for (let i = window - 1; i < values.length; i++) {
    let sum = 0;
    for (let j = i - window + 1; j <= i; j++) sum += values[j];
    result[i] = sum / window;
  }
  return result;
}

function rollingStd(values: number[], window: number): number[] {
  const means = sma(values, window);
  const result = new Array<number>(values.length).fill(NaN);
  for (let i = window - 1; i < values.length; i++) {
    let sumSq = 0;
    for (let j = i - window + 1; j <= i; j++) {
      sumSq += (values[j] - means[i]) ** 2;
    }
    result[i] = Math.sqrt(sumSq / (window - 1));
  }
  return result;
}

// EMA with adjust=False (recursive), matches pandas ewm(span, adjust=False)
function ema(values: number[], span: number): number[] {
  const k = 2 / (span + 1);
  const result = new Array<number>(values.length).fill(NaN);
  let firstValid = -1;
  for (let i = 0; i < values.length; i++) {
    if (!Number.isNaN(values[i])) {
      firstValid = i;
      break;
    }
  }
  if (firstValid === -1) return result;
  result[firstValid] = values[firstValid];
  for (let i = firstValid + 1; i < values.length; i++) {
    if (Number.isNaN(values[i])) {
      result[i] = NaN;
    } else {
      result[i] = values[i] * k + result[i - 1] * (1 - k);
    }
  }
  return result;
}

function rsiCalc(closes: number[], period = 14): number[] {
  const n = closes.length;
  const delta = new Array<number>(n).fill(NaN);
  for (let i = 1; i < n; i++) delta[i] = closes[i] - closes[i - 1];

  const gains = delta.map((d) => (Number.isNaN(d) ? NaN : Math.max(d, 0)));
  const losses = delta.map((d) => (Number.isNaN(d) ? NaN : Math.max(-d, 0)));

  const avgGains = sma(gains, period);
  const avgLosses = sma(losses, period);

  return avgGains.map((ag, i) => {
    if (Number.isNaN(ag) || Number.isNaN(avgLosses[i])) return NaN;
    const rs = avgLosses[i] === 0 ? Infinity : ag / avgLosses[i];
    return 100 - 100 / (1 + rs);
  });
}

function pctChange(values: number[], period = 1): number[] {
  const result = new Array<number>(values.length).fill(NaN);
  for (let i = period; i < values.length; i++) {
    result[i] =
      values[i - period] === 0
        ? NaN
        : (values[i] - values[i - period]) / values[i - period];
  }
  return result;
}

// ── Main ─────────────────────────────────────────────────────────────────────

export function buildFeatures(data: OHLCVRecord[]): FeatureRow[] {
  if (data.length < 60) return [];

  const closes = data.map((d) => d.close);
  const highs = data.map((d) => d.high);
  const lows = data.map((d) => d.low);
  const volumes = data.map((d) => d.volume);

  // ── Trend ───────────────────────────────────────────────────────────────
  const sma7 = sma(closes, 7);
  const sma20 = sma(closes, 20);
  const sma50 = sma(closes, 50);
  const smaRatio = sma7.map((s7, i) => s7 / sma20[i]);
  const priceToSma20 = closes.map((c, i) => c / sma20[i]);

  // ── RSI ─────────────────────────────────────────────────────────────────
  const rsiValues = rsiCalc(closes, 14);
  const rsiChange = new Array<number>(closes.length).fill(NaN);
  for (let i = 1; i < rsiValues.length; i++) {
    rsiChange[i] = rsiValues[i] - rsiValues[i - 1];
  }

  // ── MACD (normalized) ───────────────────────────────────────────────────
  const ema12 = ema(closes, 12);
  const ema26 = ema(closes, 26);
  const macd = ema12.map((e, i) =>
    Number.isNaN(e) || Number.isNaN(ema26[i]) ? NaN : e - ema26[i],
  );
  const macdSignal = ema(macd, 9);
  const macdHist = macd.map((m, i) =>
    Number.isNaN(m) || Number.isNaN(macdSignal[i]) || closes[i] === 0
      ? NaN
      : (m - macdSignal[i]) / closes[i],
  );

  // ── Bollinger Bands ─────────────────────────────────────────────────────
  const std20 = rollingStd(closes, 20);
  const bbPosition = closes.map((c, i) => {
    if (Number.isNaN(sma20[i]) || Number.isNaN(std20[i])) return NaN;
    const upper = sma20[i] + 2 * std20[i];
    const lower = sma20[i] - 2 * std20[i];
    const range = upper - lower;
    return range > 0 ? (c - lower) / range : 0.5;
  });

  // ── ATR (normalized) ───────────────────────────────────────────────────
  const trueRange = new Array<number>(closes.length).fill(NaN);
  for (let i = 1; i < closes.length; i++) {
    const hl = highs[i] - lows[i];
    const hc = Math.abs(highs[i] - closes[i - 1]);
    const lc = Math.abs(lows[i] - closes[i - 1]);
    trueRange[i] = Math.max(hl, hc, lc);
  }
  const atr = sma(trueRange, 14);
  const atrRatio = atr.map((a, i) => (closes[i] === 0 ? NaN : a / closes[i]));

  // ── Volume ──────────────────────────────────────────────────────────────
  const volumeSma20 = sma(volumes, 20);
  const volumeRatioArr = volumes.map((v, i) =>
    Number.isNaN(volumeSma20[i]) || volumeSma20[i] === 0 ? NaN : v / volumeSma20[i],
  );

  // ── OBV ─────────────────────────────────────────────────────────────────
  const obv = new Array<number>(closes.length).fill(NaN);
  obv[0] = 0;
  for (let i = 1; i < closes.length; i++) {
    const dir = closes[i] > closes[i - 1] ? 1 : closes[i] < closes[i - 1] ? -1 : 0;
    obv[i] = obv[i - 1] + volumes[i] * dir;
  }
  const obvChange = pctChange(obv);

  // ── Price changes ───────────────────────────────────────────────────────
  const priceChangeArr = pctChange(closes);
  const priceChange5d = pctChange(closes, 5);
  const highLowRange = highs.map((h, i) =>
    closes[i] === 0 ? NaN : (h - lows[i]) / closes[i],
  );

  // ── Context features ────────────────────────────────────────────────────
  const trendStrength = smaRatio.map((sr) => Math.abs(sr - 1));
  const atrRatioSma = sma(atrRatio, 20);
  const volRegime = atrRatio.map((ar, i) =>
    Number.isNaN(atrRatioSma[i]) || atrRatioSma[i] === 0 ? NaN : ar / atrRatioSma[i],
  );
  const meanRevDist = closes.map((c, i) =>
    Number.isNaN(sma50[i]) || c === 0 ? NaN : (c - sma50[i]) / c,
  );
  const rsiZone = rsiValues.map((r) => {
    if (Number.isNaN(r)) return NaN;
    return r < 30 ? 0 : r > 70 ? 2 : 1;
  });

  // Consecutive up days (capped at 5)
  const consecutiveUp = new Array<number>(closes.length).fill(NaN);
  consecutiveUp[0] = 0;
  for (let i = 1; i < closes.length; i++) {
    if (closes[i] > closes[i - 1]) {
      consecutiveUp[i] = Math.min((consecutiveUp[i - 1] || 0) + 1, 5);
    } else {
      consecutiveUp[i] = 0;
    }
  }

  // ── Build rows ────────────────────────────────────────────────────────
  const rows: FeatureRow[] = [];
  for (let i = 0; i < data.length; i++) {
    const vals = [
      smaRatio[i], priceToSma20[i], rsiValues[i], rsiChange[i],
      macdHist[i], bbPosition[i], atrRatio[i], volumeRatioArr[i],
      obvChange[i], priceChangeArr[i], priceChange5d[i], highLowRange[i],
      trendStrength[i], volRegime[i], meanRevDist[i], rsiZone[i],
      consecutiveUp[i],
    ];

    if (vals.some((v) => Number.isNaN(v) || !Number.isFinite(v))) continue;

    rows.push({
      sma_ratio: smaRatio[i],
      price_to_sma20: priceToSma20[i],
      rsi: rsiValues[i],
      rsi_change: rsiChange[i],
      macd_hist: macdHist[i],
      bb_position: bbPosition[i],
      atr_ratio: atrRatio[i],
      volume_ratio: volumeRatioArr[i],
      obv_change: obvChange[i],
      price_change: priceChangeArr[i],
      price_change_5d: priceChange5d[i],
      high_low_range: highLowRange[i],
      trend_strength: trendStrength[i],
      vol_regime: volRegime[i],
      mean_rev_dist: meanRevDist[i],
      rsi_zone: rsiZone[i],
      consecutive_up: consecutiveUp[i],
    });
  }
  return rows;
}

File: modal/features.py
import pandas as pd
import numpy as np

FEATURES = [
    "sma_ratio",        # sma_7 / sma_20 — trend direction
    "price_to_sma20",   # close / sma_20 — mean reversion
    "rsi",              # RSI(14) — momentum oscillator
    "rsi_change",       # rsi - rsi(prev) — RSI momentum
    "macd_hist",        # (macd - signal) / close — normalized MACD histogram
    "bb_position",      # (close - bb_lower) / (bb_upper - bb_lower) — Bollinger position
    "atr_ratio",        # ATR(14) / close — normalized volatility
    "volume_ratio",     # volume / volume_sma(20) — relative volume
    "obv_change",       # OBV pct change — volume-price agreement
    "price_change",     # 1-day return
    "price_change_5d",  # 5-day return
    "high_low_range",   # (high - low) / close — daily range
    # ── New context features ─────────────────────────────────────────────────
    "trend_strength",   # abs(sma_ratio - 1) — how strong the trend is
    "vol_regime",       # atr_ratio / atr_ratio_sma(20) — volatility relative to recent
    "mean_rev_dist",    # (close - sma_50) / close — distance from long-term mean
    "rsi_zone",         # 0=oversold(<30), 1=neutral, 2=overbought(>70)
    "consecutive_up",   # count of consecutive up days (capped at 5)
]

LABEL_HORIZON = 5       # predict 5 days forward
LABEL_THRESHOLD = 0.01  # 1% move


def build_features(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy()

    close = df["Close"]
    high = df["High"]
    low = df["Low"]
    volume = df["Volume"]

    # ── Trend ────────────────────────────────────────────────────────────────
    sma_7 = close.rolling(7).mean()
    sma_20 = close.rolling(20).mean()
    sma_50 = close.rolling(50).mean()
    df["sma_ratio"] = sma_7 / sma_20
    df["price_to_sma20"] = close / sma_20

    # ── RSI ──────────────────────────────────────────────────────────────────
    delta = close.diff()
    gain = delta.clip(lower=0).rolling(14).mean()
    loss = -delta.clip(upper=0).rolling(14).mean()
    rs = gain / loss
    rsi = 100 - (100 / (1 + rs))
    df["rsi"] = rsi
    df["rsi_change"] = rsi.diff()

    # ── MACD (normalized) ────────────────────────────────────────────────────
    ema_12 = close.ewm(span=12, adjust=False).mean()
    ema_26 = close.ewm(span=26, adjust=False).mean()
    macd = ema_12 - ema_26
    macd_signal = macd.ewm(span=9, adjust=False).mean()
    df["macd_hist"] = (macd - macd_signal) / close

    # ── Bollinger Bands ──────────────────────────────────────────────────────
    std_20 = close.rolling(20).std()
    bb_upper = sma_20 + 2 * std_20
    bb_lower = sma_20 - 2 * std_20
    bb_range = bb_upper - bb_lower
    df["bb_position"] = np.where(bb_range > 0, (close - bb_lower) / bb_range, 0.5)

    # ── ATR (normalized) ─────────────────────────────────────────────────────
    prev_close = close.shift(1)
    tr = pd.concat([
        high - low,
        (high - prev_close).abs(),
        (low - prev_close).abs(),
    ], axis=1).max(axis=1)
    atr = tr.rolling(14).mean()
    atr_ratio = atr / close
    df["atr_ratio"] = atr_ratio

    # ── Volume ───────────────────────────────────────────────────────────────
    volume_sma_20 = volume.rolling(20).mean()
    df["volume_ratio"] = volume / volume_sma_20

    # ── OBV ──────────────────────────────────────────────────────────────────
    direction = np.sign(close.diff())
    obv = (volume * direction).cumsum()
    df["obv_change"] = obv.pct_change()

    # ── Price changes ────────────────────────────────────────────────────────
    df["price_change"] = close.pct_change()
    df["price_change_5d"] = close.pct_change(5)
    df["high_low_range"] = (high - low) / close

    # ── Context features ─────────────────────────────────────────────────────
    df["trend_strength"] = (sma_7 / sma_20 - 1).abs()
    atr_ratio_sma = atr_ratio.rolling(20).mean()
    df["vol_regime"] = np.where(atr_ratio_sma > 0, atr_ratio / atr_ratio_sma, 1.0)
    df["mean_rev_dist"] = (close - sma_50) / close
    df["rsi_zone"] = np.where(rsi < 30, 0, np.where(rsi > 70, 2, 1)).astype(float)

    # Consecutive up days (capped at 5)
    up = (close.diff() > 0).astype(int)
    consecutive = up.copy()
    for i in range(1, len(consecutive)):
        if up.iloc[i] == 1:
            consecutive.iloc[i] = consecutive.iloc[i - 1] + 1
        else:
            consecutive.iloc[i] = 0
    df["consecutive_up"] = consecutive.clip(upper=5).astype(float)

    # ── Label: 5-day forward return > 1% ─────────────────────────────────────
    forward_return = close.shift(-LABEL_HORIZON) / close - 1
    df["label"] = (forward_return > LABEL_THRESHOLD).astype(int)

    df.replace([np.inf, -np.inf], np.nan, inplace=True)
    df.dropna(inplace=True)
    return df


def compute_balanced_accuracy(df: pd.DataFrame) -> float:
    signals = (
        (df["macd_hist"] > 0) & (df["rsi"] < 70) & (df["sma_ratio"] > 1)
    ).astype(int)
    labels = df["label"]

    true_positives = ((signals == 1) & (labels == 1)).sum()
    actual_positives = (labels == 1).sum()
    true_negatives = ((signals == 0) & (labels == 0)).sum()
    actual_negatives = (labels == 0).sum()

    sensitivity = true_positives / actual_positives if actual_positives > 0 else 0.0
    specificity = true_negatives / actual_negatives if actual_negatives > 0 else 0.0

    return round(float((sensitivity + specificity) / 2), 4)
